docs/src/coding-guidelines.md

```diff
@@ -0,0 +1,212 @@
+# Introduction
+
+This section outlines the coding standards used in the tt-mlir project. These guidelines are designed to enhance the readability and maintainability of our shared codebase. While these guidelines are not strict rules for every situation, they are essential for maintaining consistency across the repository.
+
+Our long-term aim is to have the entire codebase adhere to these conventions.
+
+Since our compiler is built on the LLVM MLIR framework, we strive to align closely with the LLVM coding style guidelines outlined here [LLVM Coding Standards](https://llvm.org/docs/CodingStandards.html).
+
+# Naming
+
+Clear and descriptive names are crucial for code readability and preventing bugs. It’s important to choose names that accurately reflect the semantics and purpose of the underlying entities, within reason. Avoid abbreviations unless they are widely recognized. Once you settle on a name, ensure consistent capitalization throughout the codebase to avoid confusion.
+
+General naming rule is to use *camel case* for most names (e.g., WorkaroundPass, isRankedTensor())
+
+1. Type Names
+    - Applies to classes, structs, enums, and typedefs.
+    - Should be nouns that describe the entity's purpose.
+    - Use **upper camel case** (e.g. TTNNOptimizerOptions, DecompositionPass).
+2. Variable Names
+    - Should be nouns, as they represent state.
+    - Use **lower camel case** (e.g. inputLayout).
+3. Function Names
+    - Represent actions and should be verb phrases
+    - Use **lower camel case** (e.g. createTTNNOptimizer(), emitTTNNAsCpp()).
+
+# Includes
+
+We prefer #includes to be listed in this order:
+
+```
+1. Main Module Header
+2. Local/Private Headers
+3. LLVM project/subproject headers (clang/..., lldb/..., llvm/..., etc)
+4. System #includes
+```
```

Comment:
Google style guidelines [recommend](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes) the opposite order (except the main module header) and I've been using it so far and forcing on PRs.

They don't elaborate their reasoning in the guidelines doc, but I've done some digging long time ago and here is the rationale:
- If you put the main module header first that means that every header in your project will be tested for missing includes in its own implementation file. That way you don't achieve any additional benefit by putting `Local/Private Headers` before other includes.
- Sometimes header files can exist without having their own implementation file and they would escape the above method, which is why some people argue that each header should have its own implementation file including it even if it is otherwise empty. 
![mqdefault-3373613200](https://github.com/user-attachments/assets/2460cf6a-88f2-4bd5-955a-bf290a0142f2)
- `Local/Private Headers` are less stable than third party includes, which are in turn less stable than standard library and system includes. It is recommended to sort includes by "stability" because for example in your local includes you might introduce some problematic redefine which might clash with some system include, and if your local includes are included first then you would get compile error in system includes, which would be very hard to debug and understand.

What do you guys think? I don't want to block the PR with this since it might be of a subjective nature, but it doesn't hurt to have a discussion :)

Reply:
I prefer the LLVM style in this case because we're working within their framework, so we keep things consistent when visiting their code. Most of the guidelines mentioned here in the guideline doc follow LLVM style, hence I would keep this one as well. This is just my personal opinion; as you mentioned, it largely comes down to individual preference.

Reply:
Oh sorry I went through their guidelines but missed the part about includes, I thought that is something that you've added just for our docs. I agree, let's be consistent with their style.